### PR TITLE
x/superfluid: ensure events are emitted and tests added

### PR DESCRIPTION
**Pull request by hieuvubk**
Closes: #2194

## Brief Changelog
 
  - *Refactor and unit test superfluid module events*
  - *Write test to make sure events emitted in right place*
  - *Update readme*

## Testing and Verifying

This change added tests and can be verified as follows:

  - *Internal events unit tests*
  - *msg server tests*

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/superfluid/README.md
Comment (p0mvn) [inline x/superfluid/README.md:490]:
```suggestion
This event is emitted in the message server after successfully creating a delegation for the given lock ID and the validator to delegate to.
```

File: x/superfluid/README.md
Comment (p0mvn) [inline x/superfluid/README.md:512]:
```suggestion
This event is emitted in the message server after undelegating the currently superfluid delegated position given by lock ID.
```

File: x/superfluid/keeper/internal/events/emit.go
Comment (p0mvn) [inline x/superfluid/keeper/internal/events/emit.go:46]: Making this comment to gather more thoughts. This does not need to be addressed.

Seems to be a common pattern. Maybe we can extract a util function that utilizes Go generics where the message being emitted is the generic type.

**Review by p0mvn — APPROVED** [review]
LGTM.

I verified that all attributes are preserved and the locations of where events are emitted are not changed.

Phenomenal work on this @hieuvubk ! and thank you for documenting it all!

File: x/superfluid/keeper/gov/suite_test.go
Comment (hieuvubk) [reply x/superfluid/keeper/gov/suite_test.go:34]: should I use this function from gamm module instead create a new function for superfluid as I did

Comment (p0mvn) [reply x/superfluid/keeper/gov/suite_test.go:34]: Eventually, we should make sure that we aren't duplicating code and using a shared function from #2186

However, for now, it is fine to have this function in superfluid. Once #2186 is merged, we can refactor

File: x/superfluid/keeper/internal/events/emit.go
Comment (p0mvn) [reply x/superfluid/keeper/internal/events/emit.go:46]: I think we can decide on this later, so going to resolve

**Comment by p0mvn** [comment]
I would like to have one more reviewer give this a look before merging since the diff is large. Overall, the change LGTM. Also, I left some a comment for follow-up work that would be good to decide on

File: x/superfluid/keeper/gov/gov_test.go
Comment (czarcas7ic) [inline x/superfluid/keeper/gov/gov_test.go:64]: Are we able to / should we break down this actions struct by item? Maybe its not a big deal but just seeing a lump of five items together doesn't make the table driven test super easy to figure out what is getting tested.

File: x/superfluid/keeper/internal/events/emit.go
Comment (czarcas7ic) [reply x/superfluid/keeper/internal/events/emit.go:46]: Agreed on this, but also not blocking as you said

File: x/superfluid/keeper/internal/events/emit_test.go
Comment (czarcas7ic) [inline x/superfluid/keeper/internal/events/emit_test.go:45]: @p0mvn is there no need to test an invalid case, such as an invalid denom?

File: x/superfluid/keeper/msg_server_test.go
Comment (czarcas7ic) [inline x/superfluid/keeper/msg_server_test.go:219]: can we be more specific just for future debugging, something like "invalid due to duration below minimum" or something along those lines?

File: x/superfluid/keeper/internal/events/emit_test.go
Comment (p0mvn) [reply x/superfluid/keeper/internal/events/emit_test.go:45]: I think this function already assumes that a denom is pre-validated somewhere else. As far as it is concerned, any received denom string is valid. So I think it should be fine without extra tests unless you would like to suggest anything specific that I'm not seeing.

File: x/superfluid/keeper/gov/gov_test.go
Comment (p0mvn) [reply x/superfluid/keeper/gov/gov_test.go:64]: Good idea, naming fields is preferable for readability IMO

File: x/superfluid/keeper/internal/events/emit_test.go
Comment (czarcas7ic) [reply x/superfluid/keeper/internal/events/emit_test.go:45]: No that all sounds good to me, was just checking!

File: x/superfluid/keeper/internal/events/emit.go
Comment (stackman27) [inline x/superfluid/keeper/internal/events/emit.go:26]: Since we are going over all the module events, what are the thoughts on using `EmitTypedEvent` instead of `EmitEvents`

It'll definitely reduce the code size and gives a better clarity. For instance the above code could be rewritten as with an event.proto file; 
```
ctx.EventManager().EmitTypedEvent(&superfluid.AssetEvent{
		denom:  denom,
	         assetType: assetType.String(), 
	})
```

ps: `EmitEvents` is labelled deprecated too

cc @alexanderbez @ValarDragon @p0mvn

Comment (p0mvn) [reply x/superfluid/keeper/internal/events/emit.go:26]: Yes, I think that's a great idea. We would have to make the event constructors exported.

For example, `newSetSuperfluidAssetEvent` -> `NewSetSuperfluidAssetEvent`

File: x/superfluid/keeper/msg_server.go
Comment (alexanderbez) [reply x/superfluid/keeper/msg_server.go:91]: Yeah we should use typed events for each message execution. We should also consider populating `MsgSuperfluidUnbondLockResponse` with relevant fields.

File: x/superfluid/keeper/internal/events/emit.go
Comment (alexanderbez) [reply x/superfluid/keeper/internal/events/emit.go:26]: Are we reusing events? I would only use constructors and separate out the event logic if we use the same ones in multiple places, otherwise it would be overkill.

File: x/superfluid/keeper/internal/events/emit.go
Comment (alexanderbez) [inline x/superfluid/keeper/internal/events/emit.go:100]: `EventManager` should never be nil, but I guess it can't hurt. Just so many LOC :-/

File: x/superfluid/README.md
Comment (alexanderbez) [inline x/superfluid/README.md:456]: I love the motivation for this, but I can see this going out of date and being difficult to maintain.

File: x/superfluid/keeper/internal/events/emit.go
Comment (hieuvubk) [reply x/superfluid/keeper/internal/events/emit.go:100]: Ahh thank you. If so, I'll remove them.

File: x/superfluid/keeper/internal/events/emit.go
Comment (ValarDragon) [reply x/superfluid/keeper/internal/events/emit.go:26]: We are in 0 rush to switch to typed events. 

I also agree that constructors seem overkill

File: x/superfluid/keeper/msg_server_test.go
Comment (stackman27) [inline x/superfluid/keeper/msg_server_test.go:199]: I noticed that events tests are very identical to their actual tests and there is repetitive code like (`TestMsgSuperfluidDelegate` and `TestMsgSuperfluidDelegate_Event`) 

is there any way we can just use one?

File: x/superfluid/keeper/msg_server_test.go
Comment (stackman27) [inline x/superfluid/keeper/msg_server_test.go:373]: this function has been defined in `app/apptesting/events.go` pretty sure we can just use that

File: x/superfluid/keeper/msg_server_test.go
Comment (stackman27) [inline x/superfluid/keeper/msg_server_test.go:246]: why do you need to pass `Suite` and `Suite.Ctx`? You can just get the context if you pass `Suite`

File: x/superfluid/keeper/msg_server_test.go
Comment (hieuvubk) [reply x/superfluid/keeper/msg_server_test.go]: Thanks for reminding

File: x/superfluid/keeper/internal/events/emit.go
Comment (stackman27) [inline x/superfluid/keeper/internal/events/emit.go:59]:
```suggestion
		sdk.NewAttribute(types.AttributeLockId, utils.Uint64ToString(lockId)),
```

lets change this for tests too

File: x/superfluid/keeper/internal/events/emit_test.go
Comment (stackman27) [inline x/superfluid/keeper/internal/events/emit_test.go:16]: Just wondering, what was the motivation behind create a new test suite? Can't we just use the `KeeperTestSuite`?

cc @p0mvn

File: x/superfluid/keeper/internal/events/emit_test.go
Comment (stackman27) [inline x/superfluid/keeper/internal/events/emit_test.go:24]: this seems to be unused

File: x/superfluid/keeper/internal/events/emit_test.go
Comment (stackman27) [inline x/superfluid/keeper/internal/events/emit_test.go:25]: this seems to be unused

Comment (hieuvubk) [reply x/superfluid/keeper/internal/events/emit_test.go]: thank you

File: x/superfluid/README.md
Comment (p0mvn) [reply x/superfluid/README.md:456]: I think we might still want to have it since these docs get auto-uploaded to the docs website. Currently, there is no info about events but it could be useful for integrators. WDYT?

**Review by ValarDragon — APPROVED** [review]
LGTM

I do have the same general point of I think this should be managed by the framework / we shouldn't do more of this, but I think this is good to merge in!

Thanks for the fix, and sorry this got blocked for so long

**Pull request merged**
